src/statement-list/statement_list_typecheck.cpp

```diff
@@ -30,87 +42,1079 @@ bool statement_list_typecheck(
+    // function body.
+    const exprt op = typecheck_identifier(tia_element, rhs.get_identifier());
+    const code_assumet assumption{op};
+    tia_element.value.add_to_operands(assumption);
+  }
+  else
+  {
+    error() << "No assignment found for assumption" << eom;
+    throw TYPECHECK_ERROR;
+  }
+}
+
+void statement_list_typecheckt::typecheck_called_tia_element(
+  const codet &op_code,
+  symbolt &tia_element)
+{
+  const symbol_exprt &call_operand = to_symbol_expr(op_code.op0());
+  const symbolt &called_function =
+    symbol_table.lookup_ref(call_operand.get_identifier());
+  const code_typet &called_type = to_code_type(called_function.type);
+  // Is it a STL function or STL function block?
+  if(
+    called_type.get(ID_statement_list_type) == ID_statement_list_function_block)
+    typecheck_called_function_block(op_code, tia_element);
+  else if(called_type.get(ID_statement_list_type) == ID_statement_list_function)
+    typecheck_called_function(op_code, tia_element);
+  else
+  {
+    error() << "Tried to call element that is no function or function block"
+            << eom;
+    throw TYPECHECK_ERROR;
+  }
+}
+
+void statement_list_typecheckt::typecheck_called_function(
+  const codet &op_code,
+  symbolt &tia_element)
+{
+  const symbol_exprt call_operand = to_symbol_expr(op_code.op0());
+  const symbolt &called_function =
+    symbol_table.lookup_ref(call_operand.get_identifier());
+  const code_typet &called_type = to_code_type(called_function.type);
+
+  // Check if function name is followed by data block.
+  if(!can_cast_expr<equal_exprt>(op_code.op1()))
+  {
+    error() << "Function calls should not address instance data blocks" << eom;
+    throw TYPECHECK_ERROR;
+  }
+
+  // Check if function interface matches the call and fill argument list.
+  const code_typet::parameterst &params = called_type.parameters();
+  code_function_callt::argumentst args;
+  std::vector<equal_exprt> assignments;
+  for(const auto &expr : op_code.operands())
+  {
+    if(can_cast_expr<equal_exprt>(expr))
+    {
+      const equal_exprt &assignment = to_equal_expr(expr);
+      assignments.push_back(assignment);
+    }
+  }
+
+  for(const auto &param : params)
+  {
+    const exprt &arg = typecheck_function_call_arguments(assignments, param);
+    args.push_back(arg);
+  }
+  const code_function_callt call{call_operand, args};
+  tia_element.value.add_to_operands(call);
+}
+
+void statement_list_typecheckt::typecheck_called_function_block(
+  const codet &op_code,
+  symbolt &tia_element)
+{
+  // TODO: Implement support for function block calls.
+  // Needs code statements which assign the parameters to the instance data
+  // block, call the function and write the result back to the parameters
+  // afterwards.
+  error() << "Calls to function blocks are not supported yet" << eom;
+  throw TYPECHECK_ERROR;
+}
+
+exprt statement_list_typecheckt::typecheck_function_call_arguments(
```

Comment:
This function does not appear to depend on any non-static members. Make `static` and remove from header, and recursively check whether any more functions could be made `static` to decrease the size of `statement_list_typecheck.h`.

Reply:
Depends on `message_handler` (see base classes of `statement_list_typecheckt`).